File: exercism/clojure/largest-series-product/src/largest_series_product.clj
(ns largest-series-product)

(defn largest-product
  ([span-length series] (cond
                          (< (count series) span-length) (throw (Exception. "span longer than series length"))
                          (and (empty? series) (not (zero? span-length)))  (throw (Exception. "Empty series and nonzero span"))
                          (and (not (empty? series)) (zero? span-length))  1
                          (and (empty? series) (zero? span-length))  1
                          (neg? span-length) (throw (Exception. "Negative span length."))
                          (true? (some (fn [c] (not (Character/isDigit c))) series))  (throw (Exception. "Non digit character"))
                          :else (largest-product span-length series 0)))
  ([span-length series current-max]
   (cond
     (empty? series)                current-max
     (< (count series) span-length) current-max
     :else                          (let [current-product (->> series
                                                               (take span-length)
                                                               (map (fn [x] (Character/getNumericValue x)))
                                                               (reduce * 1))]
                                      (largest-product span-length (rest series) (max current-product current-max))))))

File: exercism/julia/largest-series-product/largest-series-product.jl
function largest_product(str, span)
    if isempty(str) && span != 0
        throw(ArgumentError("empty string and nonzero span = $span"))
    elseif span > length(str)
        throw(ArgumentError("span = $span longer than string length $(length(str))"))
    elseif any(character -> !isdigit(character), str)
        throw(ArgumentError("invalid character in digits"))
    elseif span < 0
        throw(ArgumentError("negative span"))
    elseif span == 0
        1
    else
        # can't map accross a string, need a collect first
        biggest = 0
        digits = str |> collect |> x -> map(chr -> parse(Int, chr), x)
        for i in 1:length(digits)
            if i + span - 1 <= length(digits)
                biggest = max(biggest, prod(digits[i:i+span-1]))
            end
        end
        biggest
    end
end


